rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+
+### Additional considerations
+
+#### Database support
+
+`tsh db connect` uses raw RSA private key data to form connections. Since this cannot be supported with hardware private keys, users will instead need to use `tsh proxy db` to connect using a local proxy. Teleport Connect already uses `tsh proxy db` and will not be affected, but the WebUI may have an additional challeng to support datbase connections.
+
+#### Kubernetes support
+
+Kubernetes integration uses raw RSA private [key data to form connections](https://github.com/gravitational/teleport/blob/master/lib/kube/kubeconfig/kubeconfig.go#L164-L167). It may be possible to create a [custom auth provider plugin](https://pkg.go.dev/k8s.io/client-go@v0.24.3/tools/clientcmd/api#AuthProviderConfig) and supply it to the kubernetes Auth Info. Kubernetes support will be investigated and fixed in a follow up PR after the intial hardware private key implementation.
+
+#### Agent key support
+
+Initially, hardware private key login will not support `tsh --add-keys-to-agent`, `tsh -A`, or Proxy Recording mode, because [Adding agent keys from a hardware key](https://tools.ietf.org/id/draft-miller-ssh-agent-01.html#rfc.section.4.2.5) to a user's `ssh-agent` is [not supported in x/crypto/ssh/agent](https://github.com/golang/go/issues/16304). We can implement this support ourselves in the future.
+
+For Yubikey, users can also [manually add their keys](https://github.com/jamesog/yubikey-ssh) to their ssh-agent with `ssh-add` after logging in. However, this will not add their SSH certificate to the ssh-agent, so some additional workaround will be needed.
+
+#### PIV secret management
+
+Some PIV operations require [adminstrative access](https://developers.yubico.com/PIV/Introduction/Admin_access.html), which require one or more of the following secrets:
+
+| Name           | size     | default value                                      | function                                  |
+|----------------|----------|----------------------------------------------------|-------------------------------------------|
+| Management Key | 24 bytes | `010203040506070801020304050607080102030405060708` | private key and certificate management    |
+| PIN            | 8 chars  | `123456`                                           | sign and decrypt data, reset pin          |
+| PUK            | 8 chars  | `12345678`                                         | reset PIN when blocked by failed attempts |
+
+To simplify our implementation and limit UX impact, we will expect the user's PIV device to use the default Management Key. In the future, we may want to add support for using non-default management key to better protect the generation and retrieval of private keys on the user's PIV key, as well as PIN management if we decide to add an options like `private_key_policy: hardware_key_touch_pin`.
```

Comment:
Could we add a tsh knob? It doesn't have to a public knob.

It's a bit sad that we can't do this anymore.

Reply:
Unless we have plans to have users actively manage secrets, I don't think there's a good reason to add something like `tsh piv configure`. We can just use [ykman](https://developers.yubico.com/yubikey-manager/) for whatever functionality we need.

Reply:
What I mean is, if I change my management key from the default, could I have a flag/env var to pass the new key to tsh?

Reply:
Good point, I'll add a private flag for now.